### PR TITLE
Secure openssh-portable

### PR DESCRIPTION
**Pull request by DawsonQuick**
This branch was to secure the OpenSSH-portable to be in compliance with the Cert-C 2106 coding standard, along with fuzz testing

### CODE REVIEW

File: addrmatch.c [hunk 1]
```diff
@@ -58,6 +58,7 @@ addr_match_list(const char *addr, const char *_list)
 		return 0;
 	}
 	if ((o = list = strdup(_list)) == NULL)
+		free(o);
```
Comment (daztucker) [inline addrmatch.c:62]: this code executes when o is NULL so the free is a no-op

File: addrmatch.c [hunk 2]
```diff
@@ -118,6 +119,7 @@ addr_match_cidr_list(const char *addr, const char *_list)
...
```
Comment (daztucker) [inline addrmatch.c:124]: ditto

File: addrmatch.c [hunk 3]
```diff
@@ -14,6 +14,7 @@
  * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
  * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
  * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ * COMMENT
```
Comment (daztucker) [inline addrmatch.c:17]: no bogus changes please

File: addrmatch.c [hunk 4]
```diff
@@ -58,9 +58,10 @@ addr_match_list(const char *addr, const char *_list)
 		debug2_f("couldn't parse address %.100s", addr);
 		return 0;
 	}
-	if ((o = list = strdup(_list)) == NULL)
+	if ((o = list = strdup(_list)) == NULL) {
```
Comment (daztucker) [inline addrmatch.c:61]: this was correct until you added the unnecessary free.

File: sshbuf-getput-basic.c [hunk 5]
```diff
@@ -274,6 +274,11 @@ sshbuf_get_cstring(struct sshbuf *buf, char **valp, size_t *lenp)
 	const u_char *p, *z;
 	int r;
 
+	if(p == NULL)
+	{
+		p = 0;
+	};
```
Comment (daztucker) [inline sshbuf-getput-basic.c]: The C standard says "0"  in pointer context is a null pointer so I have no idea what you're trying to achieve with this.

File: sshbuf-getput-basic.c [hunk 6]
```diff
@@ -551,6 +556,11 @@ sshbuf_froms(struct sshbuf *buf, struct sshbuf **bufp)
 	struct sshbuf *ret;
 	int r;
 
+	if(p == NULL)
+	{
+		p = 0;
+	};
+
```
Comment (daztucker) [inline sshbuf-getput-basic.c]: ditto

File: sshbuf-misc.c [hunk 7]
```diff
@@ -78,14 +78,20 @@ sshbuf_dtob16(struct sshbuf *buf)
...
+		p = sshbuf_ptr(buf);
+	}
+
 	if (len == 0)
 		return strdup("");
 	if (SIZE_MAX / 2 <= len || (ret = malloc(len * 2 + 1)) == NULL)
 		return NULL;
-	for (i = j = 0; i < len; i++) {
+
+		for (i = j = 0; i < len; i++) {
```
Comment (daztucker) [inline sshbuf-misc.c]: no whitespace changes please, especially wrong ones like this.

File: sshbuf-misc.c [hunk 8]
```diff
@@ -279,6 +285,11 @@ sshbuf_read(int fd, struct sshbuf *buf, size_t maxlen, size_t *rlen)
 	ssize_t rr;
 	u_char *d;
 
+	if(d == NULL)
+	{
+		printf("null found");
```
Comment (daztucker) [inline sshbuf-misc.c]: you can't just send output to stdout in from the deamon as it probably doesn't have a stdout (or worse, fd 1 is used for something important).

File: sshkey.c [hunk 9]
```diff
@@ -249,6 +249,8 @@ sshkey_type_from_name(const char *name)
 	int i;
 	const struct sshkey_impl *impl;
 
+	if(impl == NULL);
```
Comment (daztucker) [inline sshkey.c]: this is a no-op and I have no idea what you're trying to do here.

File: sshkey.c [hunk 10]
```diff
@@ -316,7 +318,7 @@ sshkey_alg_list(int certs_only, int plain_only, int include_sigonly, char sep)
 	char *tmp, *ret = NULL;
 	size_t i, nlen, rlen = 0;
 	const struct sshkey_impl *impl;
-
+	if(impl == NULL);
```
Comment (daztucker) [inline sshkey.c]: ditto

File: log.c [hunk 11]
```diff
@@ -181,8 +181,9 @@ log_verbose_reset(void)
 
 	for (i = 0; i < nlog_verbose; i++)
 		free(log_verbose[i]);
-	free(log_verbose);
 	log_verbose = NULL;
+	free(log_verbose);
```
Comment (daztucker) [inline log.c:185]: free(NULL) is a no-op so it won't blow up but this adds a memory leak of log_verbose

File: ssh-keysign.c [hunk 12]
```diff
@@ -289,10 +289,11 @@ main(int argc, char **argv)
 		fatal("%s: no matching hostkey found for key %s %s", __progname,
 		    sshkey_type(key), fp ? fp : "");
 	}
-
+	if (i<5){
```
Comment (daztucker) [inline ssh-keysign.c:292]: I have no idea what this is supposed to be

File: ssh-sk-client.c [hunk 13]
```diff
@@ -389,7 +389,9 @@ sshsk_free_resident_keys(struct sshsk_resident_key **srks, size_t nsrks)
 
 	for (i = 0; i < nsrks; i++)
 		sshsk_free_resident_key(srks[i]);
+	if (srks != NULL){
```
Comment (daztucker) [inline ssh-sk-client.c:392]: free(NULL) is perfectly fine according to the C standard (see section 4.10.3.2 of ANSI C).

File: krl.c [hunk 14]
```diff
@@ -175,12 +175,10 @@ ssh_krl_free(struct ssh_krl *krl)
 	RB_FOREACH_SAFE(rb, revoked_blob_tree, &krl->revoked_keys, trb) {
 		RB_REMOVE(revoked_blob_tree, &krl->revoked_keys, rb);
 		free(rb->blob);
-		free(rb);
```
Comment (daztucker) [inline krl.c:178]: this adds a leak of rb

**Review by daztucker — COMMENTED** [review]
I don't know what the .sif files are but we don't want them.

I started reviewing this change but it seems to be composed entirely of things that are no-ops at best or outright bugs at worst and I gave up.

Assuming this is not an elaborate prank, please start from scratch and send single logical changes against the current master with a description of  the reason for the change.  "Update" is not an acceptable description.

**Comment by slshaw43** [comment]
@daztucker 

Understood and completely agree. This was apart of a senior design project for university. The sponsors/professors wanted us to make a pull request back to the master branch of openssh - portable to "attempt" to get it implemented. The coding standard they had us use was CERT-c 2016 which already was a flag to me that it would not be accepted back by you all. All of the changes were "violations" according to LDRA (static analysis tool) with CERT-C 2016 violation profile. Again, understood and thanks for at least looking over it.

have a good day/night/weekend!

Sam

**Comment by daztucker** [comment]
> The coding standard they had us use was CERT-c 2016

Is that this?
https://resources.sei.cmu.edu/downloads/secure-coding/assets/sei-cert-c-coding-standard-2016-v01.pdf

Based on the TOC those seem to be reasonable guidelines, but do not seem to match anything in this PR.

> All of the changes were "violations" according to LDRA (static analysis tool) with CERT-C 2016 violation profile.

The changes were intended to fix "violations" or create them?    If the former, something went badly wrong.  If the latter, your professors got you to deliberately send a bogus PR and waste our time?  Which university and professors?

**Comment by slshaw43** [comment]
@daztucker 

The changes were intended to fix violations that LDRA found in accordance to CERT-C 2016. After we fixed the violations, we rescanned them with LDRA and they no longer appeared to be present. This was never intended to waste anyone's time. We do apologize for the inconvenience. We have already discussed this with the professors/sponsors for the program and we will not be making anymore PRs to OpenSSH-Portable. We have already documented your concerns to ensure future teams who tackle anything like this will be more professional in committing back to the open source community (this is a one off project, no one knows if this software will be selected in the future. We will recommend to choose another software as a  #curtesy to you all). 

Again, apologies for the inconvenience, you will not see any more PR's from this team in the future.

**Pull request closed without merging**